runtime/src/main/java/org/corfudb/runtime/view/LayoutManagementView.java

```diff
@@ -212,24 +210,38 @@ public void healNode(Layout currentLayout, String endpoint) throws OutrankedExce
     }
 
     /**
-     * Attempts to merge the last 2 segments.
+     * Attempts to merge all the segments of the layout starting from the first two.
+     * Once the remaining segments can no longer be merged, attempts consensus on a new layout.
      *
      * @param currentLayout Current layout
      * @throws OutrankedException if consensus is outranked.
      */
-    public void mergeSegments(Layout currentLayout) throws OutrankedException {
-
+    public void mergeSegments(Layout currentLayout) {
         Layout newLayout;
-        if (currentLayout.getSegments().size() > 1) {
+        Predicate<Layout> shouldMergeSegments = layout -> {
+            if (layout.getSegments().size() > 1) {
+                return Sets.difference(
+                        layout.getSegments().get(1).getAllLogServers(),
+                        layout.getSegments().get(0).getAllLogServers()).isEmpty();
+            }
+            return false;
+        };
+
+        if (shouldMergeSegments.test(currentLayout)) {
 
             log.info("mergeSegments: layout is {}", currentLayout);
 
             sealEpoch(currentLayout);
 
-            LayoutBuilder layoutBuilder = new LayoutBuilder(currentLayout);
-            newLayout = layoutBuilder
-                    .mergePreviousSegment(1)
-                    .build();
+            while(shouldMergeSegments.test(currentLayout)){
+                LayoutBuilder layoutBuilder = new LayoutBuilder(currentLayout);
+                currentLayout = layoutBuilder
+                        .mergePreviousSegment(1)
```

Comment:
You've already done some good clean up from a code organization POV, but why not actually change something that has really performance impact :). Merging segments one-by-one is unnecessary, merge should merge all possible segments, otherwise there will be a lot of jitter from unnecessary seals.

Reply:
I think that's what the code is doing no? If we can merge the segments, we go to the line 230, seal the epoch on the current layout and then merge all the possible segments into a new layout until possible, then we attempt a consensus on the new layout.

Reply:
I see. Didn't read it properly.